📦 NEW: Codelab Cloudinary URL API

ahmadawais opened the pull request:
Fixes #1065

This PR adds a new codelab: `Optimize images and serve next-gen formats using Cloudinary`

After discussion with @khempenius and @traderd65 after Google I/O. I went ahead and created this new Codelab to demonstrate how the cloud can help developers avoid a build step and optimize their media both in terms of 

1. Serving images in next-gen formats, e.g., WebP and JPEG-XR.
2. Serving Efficiently encoded images with optimized size/quality.

All of that on the basis of a simple URL API. 

Things I have done while creating this codelab:

- [x] Studied the source code of other codelabs
- [x] Read the entire [Wiki](https://github.com/GoogleChrome/web.dev/wiki) guide for Web.dev
- [x] Created the Glitch embed in line with the guidelines
- [x] Kept the codelab formatting similar to other codelabs
- [x] Used the [component sandbox](https://web-dev-staging.appspot.com/sandbox/) to add relative components
- [x] Deployed the [codelab on Netlify](https://5d131654854c2b000a5f33ec--aawebdev.netlify.com/codelab-cloudinary-url-api) for a quick review

Looking forward to the review/merge. 

Peace! ✌️

ahmadawais commented:
Here's the direct staging link for the review → [https://5d131790a868e00009f0d042--web-dev-staging.netlify.com/codelab-cloudinary-url-api](https://5d131790a868e00009f0d042--web-dev-staging.netlify.com/codelab-cloudinary-url-api)

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
````diff
@@ -0,0 +1,334 @@
+https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/
+```
+
+### RESULTS
````
```suggestion
### Results
```

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,334 @@
+
+### RESULTS
+
+After replacing the link of the first image (line #25) with the fetch URL, your HTML should look like this:
```
```suggestion
After updating the `src` attribute of the first image (line #25) with the fetch URL, your HTML should look like this:
```

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,334 @@
+  the <a target="_blank" rel="noopener noreferrer" href="https://codelab-cloudinary.glitch.me/">live version</a> of your Glitch and verify that the "Efficiently encode
+  images" as well as "Serve images in next-gen formats" audits were passed.
+
+### RESULTS
```
I think the "Results" section can be omitted.

ahmadawais replied:
Should I delete this entire section?

![img](https://on.ahmda.ws/2104a7/c)

khempenius replied:
After re-reading this a bunch of times I've changed my mind. I think this would make more sense to me:

- Keep the "Results" section - but show the optimized Lighthouse result first and the unoptimized result second.
- Remove "optmized.html" so people aren't tempted to skip through the code lab just to get to the flashy Lighthouse score :)

If I were reading this I think I'd run the Lighthouse audit when I got to the last bullet point of "Lighthouse Audit". After that I think I'd expect to see the results of a successful/optimized Lighthouse run (so I could verify that I got the same results and did everything correctly). Then, if that was followed by an image of the Lighthouse report for the unoptimized "before" version, I could see the impact of what I just did, without necessarily having to run Lighthouse again.

ahmadawais replied:
@khempenius — That makes perfect sense. I've made the following changes:

- Removed `optimized.html` — good call.
- Moved the optimized results to the top.
- Removed the redundant info from the non-optimized results.

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,334 @@
+      <p>We sell the best flowers!</p>
+    </header>
+    <div class="wrapper">
+      <img src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://cloudinary-url-api.glitch.me/images/flower1.png" alt="Yellow bouquet" />
```
cloudinary-url-api -> codelab-cloudinary

ahmadawais replied:
Ah dang it. Sorry about this. Last minute glitch name change, I thought I fixed all of these.

ahmadawais replied:
🐛 FIX: 0988735

khempenius commented on src/styles/components/_asides.scss:
```diff
@@ -189,3 +189,12 @@ $KEY_POINT_COLOR: #6001ff;
 }
+
+// Colors.
+.w-c--success {
```
Remove this here & in codelab.

ahmadawais replied:
Are you sure we cannot have these new classes to add colors?

Without these classes, the table below looks quite dull and not easy to understand at a glance.

**BEFORE**:
![b](https://on.ahmda.ws/ed1d15/c)

**AFTER**:
![b](https://on.ahmda.ws/a1f2a4/c)

Also this one:


**BEFORE**:
![b](https://on.ahmda.ws/2bb543/c)

**AFTER**:
![b](https://on.ahmda.ws/af4382/c)


Let me know what you think and I'll follow your lead here.
Peace! ✌️

khempenius replied:
@robdodson is the authority here on our styles :)

ahmadawais replied:
Awesome! @robdodson what do you think? I personally think that graphs and comparison tables are inherently counter-intuitive. I always like to decrease the cognitive load for users with colors, where they can instantly understand something being wrong with `red` colored items and right with the `green` color ones. 

But hey, let me know what you think and I'll follow your lead here. 

Looking forward, peace! ✌️

mfriesenhahn replied:
I'm inclined to agree with Katie. We're trying to avoid custom styles in general because, when needed, we want to be able to update site styles with confidence that nothing will break.

ahmadawais replied:
@mfriesenhahn makes sense. I am all for that. I just need a way to make it easier for the user reading the Codelab to be easily able to differentiate between the good and bad results — by way of red and green colors. 

I was unable to find a class that just changes the colors to the red/green colors defined for web.dev. Are there such classes available that I'd have missed?

Or do you think I should just remove the colors from both the SCSS and the markdown file? 👀

robdodson replied:
I don't think we have specific styles for this kind of UI so I think I'd suggest removing them.

ahmadawais replied:
Sure, removed in 22aed01

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,334 @@
+    </header>
+    <div class="wrapper">
+      <img src="https://res.cloudinary.com/demo/image/fetch/q_auto,f_auto/https://cloudinary-url-api.glitch.me/images/flower1.png" alt="Yellow bouquet" />
+      <div class="price">Yellow bouquet - $9</div>
```
I think I'd make the glitch slightly shorter (maybe only 3 or 4 flowers?). This cuts down on the amount of changes people need to make the HTML & makes this code snippet shorter too.

ahmadawais replied:
👌 IMPROVE: Shorten to Three images → d4040af

ahmadawais commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,334 @@
+description: |
+  In this codelab, learn how to use Cloudinary with its URL API to optimize JPEG and
+  PNG images for faster download and serve next-gen formats.
+glitch: codelab-cloudinary?path=index.html:25:0
```
I got the codelab reviewed at a local meetup. 8 out of 11 attendees had problems figuring out how to open `index.html` since it was there first time with Glitch and small screens glitch hides the left Files Explorer. 

Then I worked around a way for the codelab to open up with the exact file instead `index.html`. This helped them get started quickly.

So, I'd like to request to keep `?path=index.html:25:0` part there for better Developer Experience.

Thoughts?

khempenius replied:
That sounds reasonable to me.

@robdodson - Can you confirm that adding URL params to the glitch name won't break anything?

robdodson replied:
I think you can use the path param in YAML to do this:

```
glitch: codelab-cloudinary
path: index.html:25:0
```

ahmadawais replied:
@robdodson thanks for sharing that. Fixed in db09fa3

ahmadawais commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,334 @@
+
+<figure class="w-figure">
+  <a href="https://cloudinary.com/users/register/free" target="_blank" rel="noopener noreferrer">
+    <img class="w-screenshot" src="./cloudname.gif" alt="Setting a custom Cloud Name" />
```
On the meetup, I found out that many users would skip looking/setting up a custom `cloudName`. It's something we are working on at Cloudinary. But once I added this GIF it made them set it up easily.

I have optimized this GIF to be smaller in size than other images. It's 76 KB only. So, my question is can we keep it? 

Let me know and I'm game either way!

mfriesenhahn replied:
As a rule, we favor video over GIFs: https://web.dev/handbook/markup-media/#video-hosted-on-web.dev

ahmadawais replied:
@mfriesenhahn Sure thing. I have uploaded the [video here](https://github.com/GoogleChrome/web.dev/blob/0ba6faf3e7d73595d4a95727a7de326701e673b3/src/site/content/en/fast/codelab-cloudinary-url-api/coudname.mp4). 

>Once you have a video ready, message your content reviewer, and they'll help you upload it to the web.dev Google Cloud Storage bucket.

I hope you can take care of this now. :)

ahmadawais commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,334 @@
+
+## ✔︎ Check-in
+
+Your `index.html` file should now look like this:
```
Sure thing!

ahmadawais replied:
🐛 FIX: in 399812f

ahmadawais commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,334 @@
+
+There are impressive web performance gains by using <a href="https://cloudinary.com/documentation/image_optimization">cloud to optimize images</a> without losing quality.
+
+<div class="w-table-wrapper">
```
Glad you liked it. Fun comparison. 

P.S. I use [Numi](https://numi.io/) mac calculator to calculate the optimized image sizes.

P.P.S In case you're into this, I've written a custom extension for Numi calculator where I only have to write `optim(newSize;oldSize)` of an image to calculate the percentage savings.

![optim](https://on.ahmda.ws/29b9a9/c)

Here's the code ↓

```js
/**
 * Optimization percentage in new vs old size.
 *
 * @usage optim(newSize;oldSize)
 * @usage optim(10;100)
 * @param Number newSize The new reduced size.
 * @param Number oldSize The old large size.
 * @author Ahmad Awais <https://twitter.com/MrAhmadAwais>
 */

numi.addFunction({ id: 'optim', phrases: 'optim' }, function(values) {
	return {
		double: (values[0].double / values[1].double) * 100 - 100
	};
});
```

ahmadawais commented:
@khempenius 

>This is looking good! I left some comments - but it's all pretty minor stuff.

Super glad you liked it. 😇

Thanks a lot for the quick review. I have resolved most of the comments and made relevant improvements + fixes. There are three things left open, I have left comments for you to review.

Looking forward to your thoughts and I'll address them after that.

Peace! ✌️

ahmadawais commented:
Latest build URL for reference. 

→ https://deploy-preview-1041--web-dev-staging.netlify.com//codelab-cloudinary-url-api/

ahmadawais commented:
@khempenius I've resolved all the comments, can you please review it again?

→ Deploy preview: [codelab-cloudinary-url-api](https://deploy-preview-1041--web-dev-staging.netlify.com/)

Looking forward, peace! ✌️

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
````diff
@@ -0,0 +1,276 @@
+
+```
+
+Which means that you can auto-optimize any image by prepending the following to its URL.
````
```suggestion
This works for other images too. You can auto-optimize any image by prepending the following to its URL:
```

ahmadawais commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,276 @@
+
+## The Cloudinary Fetch URL
+
+This codelab demonstrates how you can use the power of a cloud to optimize your images and serve next-gen formats dynamically. You will learn how to use dynamic cloud Fetch and Image Transformations features. You will replace the image links with Fetch URLs (explained below) that will automagically upload, transform, and optimize your images.
```
Done in 761ec21.

ahmadawais commented:
@khempenius Resolved and fixed all your comments. Looking forward to getting this PR merged.

→ Deploy preview: [codelab-cloudinary-url-api](https://deploy-preview-1041--web-dev-staging.netlify.com/codelab-cloudinary-url-api/)

Peace! ✌️

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,274 @@
+
+### Create a Fetch URL
+
+To create a fetch URL, you have to prepend the following prefix to the existing URL of the image.
```
```suggestion
To create a fetch URL, prepend the following prefix to the existing URL of the image:
```

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,274 @@
+
+To create a fetch URL, you have to prepend the following prefix to the existing URL of the image.
+
+An example fetch URL looks like this:
```
```suggestion
```

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
````diff
@@ -0,0 +1,274 @@
+</div>
+```
+
+This helps save more than 90% on image size.
````
```suggestion
This decreases image size by more than 90%.
```

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,274 @@
+  </figcaption>
+</figure>
+
+If you've optimized your images correctly the Lighthouse audit should have a perfect score. As you can see the page load time is down by 4.05 secs. That is impressive.
```
```suggestion
If you've optimized your images correctly the Lighthouse audit should have a perfect score. As you can see First Meaningful Paint is down by 2.5 secs. That is impressive.
```

khempenius replied:
(Changing this to match the image, which highlights First Meaningful Paint.)

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,274 @@
+
+### Results
+
+Lighthouse audit results for when images are optimized.
```
```suggestion
```

khempenius replied:
Captions should be sufficient.

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,274 @@
+
+If you've optimized your images correctly the Lighthouse audit should have a perfect score. As you can see the page load time is down by 4.05 secs. That is impressive.
+
+_Lighthouse audit results when images are not optimized._
```
```suggestion
```

khempenius commented on src/site/content/en/fast/codelab-cloudinary-url-api/index.md:
```diff
@@ -0,0 +1,274 @@
+
+## Image Performance Gains
+
+There are impressive web performance gains by using <a href="https://cloudinary.com/documentation/image_optimization">cloud to optimize images</a> without losing quality.
```
```suggestion
These are the results of using <a href="https://cloudinary.com/documentation/image_optimization">cloud to optimize images</a> without losing quality.
```

mfriesenhahn commented:
@khempenius, since this post won't be published until the Thumbor piece is ready, I've moved it to the next sprint. I may not get to the review for a bit to focus on stuff that's going out earlier. Just wanted to give you a heads-up!

ahmadawais commented:
@khempenius Resolved all the comments. Thanks a lot for reviewing this and making it near perfect.

@mfriesenhahn Looking forward to the merged PR. Let me know if there's anything I could help you with on this.

Peace! ✌️

ahmadawais commented:
🐛 FIXED: Merge conflict in 127180b.

ahmadawais commented:
@mfriesenhahn do you need anything from me to merge this PR? I've addressed all the issues raised and resolved merge conflicts a couple of times.

Looking forward, peace! ✌️

mfriesenhahn commented:
> @mfriesenhahn do you need anything from me to merge this PR? I've addressed all the issues raised and resolved merge conflicts a couple of times.

Hey, @ahmadawais, I added a comment last week: 
> @khempenius, since this post won't be published until the Thumbor piece is ready, I've moved it to the next sprint. I may not get to the review for a bit to focus on stuff that's going out earlier. Just wanted to give you a heads-up!

So, I'll review this post and the Thumbor one together during the next sprint, which starts 7/17--unless there's a reason this needs to go out sooner?

khempenius commented:
Hi @ahmadawais - I just talked with @mfriesenhahn about this. He's got a lot on his plate right now for this sprint, so he was planning to push his review to the July 17th sprint. (We just started a new sprint system for web.dev and are still working out the kinks.) Depending on how other things go, he _might_ be able to get around to it before then. Thank you for your patience!

ahmadawais commented:
@mfriesenhahn @khempenius thanks for letting me know. I was not aware of the sprints date (figured it was internal talk) so asked it anyway. Looking forward to getting this PR merged in the next sprint. Thanks for your time.

Peace! ✌️

ahmadawais commented:
@mfriesenhahn @khempenius just a friendly reminder about this PR. :)

I am not sure generally how long are the sprints at Google, it's been about a week since 17th, so sending in a polite reminder. 

Peace! ✌️

mfriesenhahn commented:
> @mfriesenhahn @khempenius just a friendly reminder about this PR. :)
> 
> I am not sure generally how long are the sprints at Google, it's been about a week since 17th, so sending in a polite reminder.
> 
> Peace! ✌️

Thanks, @ahmadawais. To keep this moving, I've asked another tech writer, @katejeffreys, to review. This sprint ends 7/30, so this codelab will go out by then!

katejeffreys commented:
I made a PR into Ahmad's branch with my suggested edits. https://github.com/ahmadawais/web.dev/pull/1 

@ahmadawais @khempenius @mfriesenhahn thanks for all your hard work on this so far. I got most of the way through editing this when I thought to look at the commit history and saw that it had already been pretty well thrashed out by the three of you. :/

I thought I would finish mine anyway in case it's helpful. Sorry if these edits are repetitive or not what you were going for!

Staged here: https://20190723t174210-dot-web-dev-staging.appspot.com

ahmadawais commented:
Thank you, @katejeffreys for taking time out and reviewing this codelab. I've gone ahead:

- 📦 Merged the changes you sent as a PR. 
- 🐛 Fixed the merge conflicts.
- 👌 The codelab now [looks like this →](https://deploy-preview-1041--web-dev-staging.netlify.com/codelab-cloudinary-url-api/)

@mfriesenhahn All ready for your review. Looking forward, peace! ✌️

mfriesenhahn commented:
If @katejeffreys is happy, I'm happy! Kate, just let me know if this is ready to merge :)

katejeffreys commented:
I think it's good to go. Thanks all

ahmadawais commented:
@robdodson @mfriesenhahn thank you for approving & merging the codelab.

@khempenius I see that the codelab is now live at [https://web.dev/codelab-cloudinary-url-api/](https://web.dev/codelab-cloudinary-url-api/) but it's not connected with this [Imagemin article](https://web.dev/use-imagemin-to-compress-images/) — basically, there's no way to access the codelab unless you have the exact URL.

I also searched for it in the search bar and no dice. Can I do something here to make that happen?

I remember adding [this line](https://github.com/GoogleChrome/web.dev/pull/1041/files#diff-aa009299106d6b7b9a3e162cd4ba5d10R11) but for some reason, it didn't work.

Looking forward, peace! ✌️

mfriesenhahn commented:
Hi, @ahmadawais! We're planning to push all our image CDN content at one time, including an introductory post about image CDNs, which will link to this codelab and others. We're targeting next week!

robdodson commented:
yep, the post is currently listed as a "draft" which means it does get staged but it's not crawlable and nothing links to it. As Michael mentioned, these should be going live soon and we'll ping you when it's up.
